File: C++/PACKING.cpp
// backtracking을 이용해서 답을 찾는 방법을 잘 기억해 두자.
// 그렇게 어려운 문제는 아니었다.

#include <iostream>
#include <cstring>
#include <vector>

using namespace std;

int memo[101][1002];

int no_of_items;
typedef struct item {
    string name;
    int volume, desp;
} item;
item items[101];

int get_max_desp(int, int);
void reconstruct(int, int, vector<string>&);

int main() {
    int i, j, k, cap;
    for(scanf("%d", &i), j = 0; j < i; j++){
        vector<string> picked;
        memset(memo, -1, sizeof(memo));
        scanf("%d %d", &no_of_items, &cap);
        for(k=0; k<no_of_items; k++){
            cin >> items[k].name >> items[k].volume >> items[k].desp; 
        }
        reconstruct(0, cap, picked);
        printf("%d %d\n", memo[0][cap], (int)picked.size());
        for(k=0; k<picked.size(); k++){
            cout << picked[k] << endl;
        }
    }
}

int get_max_desp(int s_index, int cap){
    int &result = memo[s_index][cap];
    if(result != -1){
        return result;
    }
    if(s_index >= no_of_items || cap <=0)
        return result = 0;
    
    result = get_max_desp(s_index + 1, cap);
    if(items[s_index].volume <= cap){
        int tmp = items[s_index].desp + get_max_desp(s_index + 1, cap - items[s_index].volume);
        result = result > tmp ? result : tmp; 
    }
    return result;
}

void reconstruct(int s_index, int cap, vector<string> &picked){
    if(s_index >= no_of_items || cap <= 0)
        return;
    if(get_max_desp(s_index,cap) == get_max_desp(s_index + 1,cap)){
        reconstruct(s_index + 1, cap, picked);
    }
    else
    {
        picked.push_back(items[s_index].name);
        reconstruct(s_index+1, cap - items[s_index].volume, picked);
    }
}

File: Python/PACKING.py
## 알고리즘은 맞는데, 제한시간 초과!!
## memo를 dictionary에서 list로 바꾸니까 해결됨!
## 아무래도 dictionary구조로 동적으로 메모를 하는 게 느린듯?
## backtracking을 잘 기억해 두자!

import sys

memo = []
names = []
volumes = []
need = []
picked = []
sys.setrecursionlimit(10**6)

def get_max_need(s_index, cap):
    result = (0, [])
    if memo[s_index][cap] != -1:
        return memo[s_index][cap]
    if s_index >= len(names) or cap <=0:
        return 0
    result = get_max_need(s_index + 1, cap)
    if volumes[s_index] <= cap:
        result = max(result, need[s_index] + get_max_need(s_index + 1, cap - volumes[s_index]))
    memo[s_index][cap] = result
    return result

def reconstruct(s_index, cap):
    if s_index >= len(names):
        return
    if get_max_need(s_index, cap) == get_max_need(s_index + 1, cap):
        reconstruct(s_index + 1, cap)
    else:
        picked.append(names[s_index])
        reconstruct(s_index+1, cap - volumes[s_index])

for _ in range(int(input())):
    # initialization
    maxv = 0
    names = []
    volumes = []
    need = []

    no_of_items, cap = map(int, sys.stdin.readline().strip().split())
    memo = [[-1] * (cap + 1) for _ in range(no_of_items + 1)]
    for _ in range(no_of_items):
        name, v, n = input().strip().split();
        names.append(name)
        volumes.append(int(v))
        need.append(int(n))
    picked = []
    reconstruct(0, cap)
    print(memo[0][cap], len(picked))
    for item in picked:
        print(item)

        